### PR TITLE
PSF measurement, simulation, and deconvolution

### PR DESCRIPTION
**Pull request by talonchandler**
- paired with [this confluence page](https://czbiohub.atlassian.net/wiki/spaces/dashboard/pages/3411214430/2024-02-05+modeling+mantis+fluorescence+PSFs) and [this confluence page](https://czbiohub.atlassian.net/wiki/spaces/dashboard/pages/3567288346/2024-05-05+PSF+fitting+for+mantis+deconvolution) 
- 2024-05-07 - now includes draft CLI calls `psf-from-beads` which estimates an averaged PSF from a bead volume, and `deconvolve` which uses the estimated PSF to deconvolve

TODO:
 - [ ] update pre-commit

### CODE REVIEW

File: mantis/cli/deconvolve.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+    # Check if scales match
+    if psf_scale != zyx_scale:
+        click.echo(
+            f"Warning: PSF scale {psf_scale} does not match data scale {zyx_scale}. "
+            "Consider resampling the PSF."
+        )
+
+    for t in range(1):#T):
+        for c in range(C):
```
Comment (talonchandler) [inline mantis/cli/deconvolve.py:60]: Kludge here...needs multiprocessing and `process_single_position`

File: mantis/cli/characterize.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+def characterize(
+    input_position_dirpaths: List[str],
+    config_filepath: str,
+    output_dirpath: str,
+):
+    """
+    Characterize the point spread function (PSF) from bead images in an html report
+
+    >> mantis characterize -i ./beads.zarr/*/*/* -c ./characterize_params.yml -o ./
+    """
```
Comment (talonchandler) [inline mantis/cli/characterize.py:33]: Only characterizes first position. Could use better docs.

**Comment by ieivanov** [comment]
@talonchandler would you like to keep `scripts/fit_psf_to_beads.py`, `scripts/simple_psf.py`, and `scripts/simulate_psf.py`? They may need to have their deskew updated. These scripts call on `deskew/_deskew_matrix` which I think now is deprecated. Or is `deskew/_get_transform_matrix` the deprecated one? `deskew_data` uses
```
    matrix = _get_transform_matrix(
        raw_data.shape, ls_angle_deg, px_to_scan_ratio, keep_overhang
    )
```
but I think that may be a bug? I'd like to remove one of `_deskew_matrix` or `_get_transform_matrix`.

**Comment by ieivanov** [comment]
P. S. If that's a bug in the deskew, let's fix it in a separate PR that we can merge right away

**Comment by ieivanov** [comment]
> P. S. If that's a bug in the deskew, let's fix it in a separate PR that we can merge right away

This is a bug in this branch only, I'll fix it

**Comment by talonchandler** [comment]
I removed `scripts/fit_psf_to_beads.py`, `scripts/simple_psf.py`, and `scripts/simulate_psf.py` because I don't think we'll need them. I can resurrect them if that changes. 

Those are the only places that `_deskew_matrix` is used, so I've deleted that function, too. 

With those deletions we can focus our attention on the `_get_transform_matrix`, which was regressing from the `main` branch (the GPU deskew branch changed the affine transformation matrix). I added a test to catch this problem, then fixed the problem. 

I spot-tested a deskew on real data, so I now think the deskew in this branch is ready (a minor refactor from `main`)

@ieivanov thanks for flagging this. I just took a look, and I think these `deskew` functions were the only places where accidental regressions happened.

**Comment by ieivanov** [comment]
I'll test the refactored `measure_psf` today and then we can merge

**Review by ieivanov — APPROVED** [review]
I've tested `measure_psf` on mantis, I think this PR is ready to merge

File: mantis/cli/characterize_psf.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+@output_dirpath()
+def characterize_psf(
+    input_position_dirpaths: List[str],
+    config_filepath: str,
+    output_dirpath: str,
+):
+    """
+    Characterize the point spread function (PSF) from bead images and output an html report
+
+    >> mantis characterize -i ./beads.zarr/*/*/* -c ./characterize_params.yml -o ./
```
Comment (talonchandler) [inline mantis/cli/characterize_psf.py:96]:
```suggestion
    >> mantis characterize-psf -i ./beads.zarr/*/*/* -c ./characterize_params.yml -o ./
```

File: mantis/cli/estimate_psf.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+@output_dirpath()
+def estimate_psf(
+    input_position_dirpaths: List[str],
+    config_filepath: str,
+    output_dirpath: str,
+):
+    """
+    Estimate the point spread function (PSF) from bead images
+
+    >> mantis psf_from_beads -i ./beads.zarr/*/*/* -c ./psf_params.yml -o ./psf.zarr
```
Comment (talonchandler) [inline mantis/cli/estimate_psf.py:32]:
```suggestion
    >> mantis estimate-psf -i ./beads.zarr/*/*/* -c ./psf_params.yml -o ./psf.zarr
```

**Comment by talonchandler** [comment]
LGTM!

**Pull request merged**